File: ChatGPT(예진)/Python (수정 완료)/Transmission_Algorithms_Hamming_CRC_RSA.py
import random

# 해밍 코드 구현
def generate_hamming_code(data):
    n = len(data)
    m = 0
    while 2 ** m < n + m + 1:
        m += 1
    hamming_code = [None] * (n + m)
    j = 0
    for i in range(n + m):
        if i + 1 == 2 ** j:
            hamming_code[i] = 0
            j += 1
        else:
            if j < n:  # j가 데이터의 길이를 벗어나지 않도록 확인
                hamming_code[i] = int(data[j])
                j += 1
            else:
                hamming_code[i] = 0  # 데이터 길이를 벗어나면 0으로 채움
    for i in range(m):
        p = 2 ** i - 1
        count = 0
        for j in range(p, n + m, 2 * (p + 1)):
            for k in range(j, min(j + p + 1, n + m)):
                if hamming_code[k] == 1:
                    count += 1
        if count % 2 == 0:
            hamming_code[p] = 0
        else:
            hamming_code[p] = 1
    return ''.join(map(str, hamming_code))

# CRC 코드 구현
def generate_crc_code(data, divisor):
    crc = list(data)
    divisor_length = len(divisor)
    for i in range(len(data) - divisor_length + 1):
        if crc[i] == '1':
            for j in range(divisor_length):
                crc[i + j] = str(int(crc[i + j]) ^ int(divisor[j]))
    return ''.join(crc)

# RSA 암호화 알고리즘 구현
def gcd(a, b):
    while b != 0:
        a, b = b, a % b
    return a

def mod_inverse(a, m):
    m0, x0, x1 = m, 0, 1
    while a > 1:
        q = a // m
        m, a = a % m, m
        x0, x1 = x1 - q * x0, x0
    return x1 + m0 if x1 < 0 else x1

def generate_rsa_keypair(p, q):
    n = p * q
    phi = (p - 1) * (q - 1)
    e = random.randint(2, phi - 1)
    while gcd(e, phi) != 1:
        e = random.randint(2, phi - 1)
    d = mod_inverse(e, phi)
    return ((e, n), (d, n))

# 예제 데이터 생성
data = "101010"
divisor = "1101"
p = 61
q = 53

# 해밍 코드 결과 출력
print("Hamming Code:")
hamming_code = generate_hamming_code(data)
print("Original Data:", data)
print("Hamming Code:", hamming_code)

# CRC 코드 결과 출력
print("\nCRC Code:")
crc_code = generate_crc_code(data + "0" * (len(divisor) - 1), divisor)
print("Original Data:", data)
print("CRC Code:", crc_code)

# RSA 암호화 알고리즘 결과 출력
print("\nRSA Encryption:")
public_key, private_key = generate_rsa_keypair(p, q)
plaintext = 42
encrypted_text = pow(plaintext, public_key[0], public_key[1])
print("Plaintext:", plaintext)
print("Encrypted Text:", encrypted_text)

# RSA 복호화 알고리즘 결과 출력
print("\nRSA Decryption:")
decrypted_text = pow(encrypted_text, private_key[0], private_key[1])
print("Decrypted Text:", decrypted_text)

File: ChatGPT(예진)/Java (수정 완료)/Transmission_Algorithms_Hamming_CRC_RSA.java
import java.util.Random;

public class Transmission_Algorithms_Hamming_CRC_RSA {
    // Hamming code implementation
    public static String generateHammingCode(String data) {
        int n = data.length();
        int m = 0;
        while (Math.pow(2, m) < n + m + 1) {
            m++;
        }
        char[] hammingCode = new char[n + m];
        int j = 0;
        for (int i = 0; i < n + m; i++) {
            if (i + 1 == Math.pow(2, j)) {
                hammingCode[i] = '0';
                j++;
            } else {
                if (j < n) {
                    hammingCode[i] = data.charAt(j);
                    j++;
                } else {
                    hammingCode[i] = '0';
                }
            }
        }
        for (int k = 0; k < m; k++) {
            int p = (int) Math.pow(2, k) - 1;
            int count = 0;
            for (int l = p; l < n + m; l += 2 * (p + 1)) {
                for (int o = l; o < Math.min(l + p + 1, n + m); o++) {
                    if (hammingCode[o] == '1') {
                        count++;
                    }
                }
            }
            if (count % 2 == 0) {
                hammingCode[p] = '0';
            } else {
                hammingCode[p] = '1';
            }
        }
        return new String(hammingCode);
    }

    // CRC code implementation
    public static String generateCrcCode(String data, String divisor) {
        char[] crc = data.toCharArray();
        int divisorLength = divisor.length();
        for (int i = 0; i < data.length() - divisorLength + 1; i++) {
            if (crc[i] == '1') {
                for (int j = 0; j < divisorLength; j++) {
                    crc[i + j] = (char) (((crc[i + j] - '0') ^ (divisor.charAt(j) - '0')) + '0');
                }
            }
        }
        return new String(crc);
    }

    // RSA encryption algorithm implementation
    public static int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    public static int modInverse(int a, int m) {
        int m0 = m, x0 = 0, x1 = 1;
        while (a > 1) {
            int q = a / m;
            int temp = m;
            m = a % m;
            a = temp;
            temp = x0;
            x0 = x1 - q * x0;
            x1 = temp;
        }
        return x1 < 0 ? x1 + m0 : x1;
    }

    public static int[] generateRsaKeyPair(int p, int q) {
        int n = p * q;
        int phi = (p - 1) * (q - 1);
        Random random = new Random();
        int e = random.nextInt(phi - 2) + 2;
        while (gcd(e, phi) != 1) {
            e = random.nextInt(phi - 2) + 2;
        }
        int d = modInverse(e, phi);
        return new int[]{e, n, d};
    }

    public static void main(String[] args) {
        // Example data creation
        String data = "101010";
        String divisor = "1101";
        int p = 61;
        int q = 53;

        // Hamming code result output
        System.out.println("Hamming Code:");
        String hammingCode = generateHammingCode(data);
        System.out.println("Original Data: " + data);
        System.out.println("Hamming Code: " + hammingCode);

        // CRC code result output
        System.out.println("\nCRC Code:");
        String crcCode = generateCrcCode(data + "0".repeat(divisor.length() - 1), divisor);
        System.out.println("Original Data: " + data);
        System.out.println("CRC Code: " + crcCode);

        // RSA encryption algorithm result output
        System.out.println("\nRSA Encryption:");
        int[] keyPair = generateRsaKeyPair(p, q);
        int plaintext = 42;
        int encryptedText = (int) Math.pow(plaintext, keyPair[0]) % keyPair[1];
        System.out.println("Plaintext: " + plaintext);
        System.out.println("Encrypted Text: " + encryptedText);

        // RSA decryption algorithm result output
        System.out.println("\nRSA Decryption:");
        int decryptedText = (int) Math.pow(encryptedText, keyPair[2]) % keyPair[1];
        System.out.println("Decrypted Text: " + decryptedText);
    }
}
